includes/Export.php

```diff
@@ -300,44 +370,115 @@ protected function dumpFrom( $cond = '' ) {
 				$prev = $this->db->bufferResults( false );
 			}
 
-			Hooks::run( 'ModifyExportQuery',
+			$result = null; // Assuring $result is not undefined, if exception occurs early
+			try {
+				Hooks::run( 'ModifyExportQuery',
 						array( $this->db, &$tables, &$cond, &$opts, &$join ) );
 
-			# Do the query!
-			$result = $this->db->select( $tables, '*', $cond, __METHOD__, $opts, $join );
-			$wrapper = $this->db->resultObject( $result );
-			# Output dump results
-			$this->outputPageStream( $wrapper );
-			if ( $this->list_authors ) {
-				$this->outputPageStream( $wrapper );
+				# Do the query!
+				$result = $this->selectRows( $tables, '*', $cond, __METHOD__, $opts, $join );
+				# Output dump results
+				$this->outputPageStream( $result );
+
+				if ( $this->buffer == WikiExporter::STREAM ) {
+					$this->db->bufferResults( $prev );
+				}
+			} catch ( Exception $e ) {
+				// Throwing the exception does not reliably free the resultset, and
+				// would also leave the connection in unbuffered mode.
+
+				// Freeing result
+				try {
+					if ( $result instanceof ResultWrapper ) {
+						$result->free();
+					}
+				} catch ( Exception $e2 ) {
+					// Already in panic mode -> ignoring $e2 as $e has
+					// higher priority
+				}
+
+				// Putting database back in previous buffer mode
+				try {
+					if ( $this->buffer == WikiExporter::STREAM ) {
+						$this->db->bufferResults( $prev );
+					}
+				} catch ( Exception $e2 ) {
+					// Already in panic mode -> ignoring $e2 as $e has
+					// higher priority
+				}
+
+				// Inform caller about problem
+				throw $e;
 			}
+		}
+	}
 
-			if ( $this->buffer == WikiExporter::STREAM ) {
-				$this->db->bufferResults( $prev );
+	/**
+	 * This method makes the main database query to fetch all rows required to generate a dump.
+	 *
+	 * When $cond argument is empty, we assume that we're generating a full dump. Instead of using unbuffered query,
+	 * let's just send a set of small batched queries and yield each row returned. outputPageStream method
+	 * uses foreach to consume the passed $result object.
+	 *
+	 * @param $tables
+	 * @param $vars
+	 * @param $cond
+	 * @param $fname
+	 * @param $opts
+	 * @param $join
+	 * @return ResultWrapper|Generator
+	 */
+	private function selectRows($tables, $vars, $cond, $fname, $opts, $join) {
+		// we're not doing a full dump, let's just make a single query
+		if ( !empty( $cond ) ) {
+			return $this->db->select( $tables, $vars, $cond, $fname, $opts, $join );
+		}
+		else {
+			// as said above, we do want to make a set of small, buffered queries
+			$this->db->bufferResults( true );
+
+			$page_max = (int) $this->db->selectField( 'page','MAX(page_id)', '', $fname );
+
+			// fetch 500 pages in each batch
+			$offset_start = 1;
+			$batch_size = 500; # this returns up to ~100k rows on yugioh
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Move this trailing comment on the previous empty line. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS139)